apps/bilateral_grid/Makefile

```diff
@@ -1,57 +1,53 @@
 include ../support/Makefile.inc
 
-all: $(BIN)/filter
+all: $(BIN)/$(HL_TARGET)/filter
 
-$(BIN)/bilateral_grid.generator: bilateral_grid_generator.cpp $(GENERATOR_DEPS)
+$(GENERATOR_BIN)/bilateral_grid.generator: bilateral_grid_generator.cpp $(GENERATOR_DEPS)
 	@mkdir -p $(@D)
 	$(CXX) $(CXXFLAGS) -g $(filter-out %.h,$^) -o $@ $(LDFLAGS) $(HALIDE_SYSTEM_LIBS)
 
-$(BIN)/bilateral_grid.a: $(BIN)/bilateral_grid.generator
+$(BIN)/%/bilateral_grid.a: $(GENERATOR_BIN)/bilateral_grid.generator
 	@mkdir -p $(@D)
-	$^ -g bilateral_grid -o $(BIN) -f bilateral_grid target=$(HL_TARGET) auto_schedule=false
+	$^ -g bilateral_grid -o $(@D) -f bilateral_grid target=$* auto_schedule=false
 
-$(BIN)/bilateral_grid_auto_schedule.a: $(BIN)/bilateral_grid.generator
+$(BIN)/%/bilateral_grid_auto_schedule.a: $(GENERATOR_BIN)/bilateral_grid.generator
 	@mkdir -p $(@D)
-	$^ -g bilateral_grid -o $(BIN) -f bilateral_grid_auto_schedule target=$(HL_TARGET)-no_runtime auto_schedule=true -e static_library,h,schedule
+	$^ -g bilateral_grid -o $(@D) -f bilateral_grid_auto_schedule target=$*-no_runtime auto_schedule=true -e static_library,h,schedule
 
-$(BIN)/viz/bilateral_grid.a: $(BIN)/bilateral_grid.generator
+$(BIN)/%/filter: $(BIN)/%/bilateral_grid.a $(BIN)/%/bilateral_grid_auto_schedule.a filter.cpp
 	@mkdir -p $(@D)
-	$^ -g bilateral_grid -o $(BIN)/viz target=$(HL_TARGET)-trace_all
+	$(CXX) $(CXXFLAGS) -ffast-math -Wall -Werror -I$(BIN)/$* $^ -o $@ $(IMAGE_IO_FLAGS) $(LDFLAGS)
 
-$(BIN)/filter: $(BIN)/bilateral_grid.a $(BIN)/bilateral_grid_auto_schedule.a filter.cpp
+$(BIN)/%/filter_viz: $(BIN)/%-trace_all/bilateral_grid.a filter.cpp ../../bin/HalideTraceViz
```

Comment:
Could we just delete this rule entirely, and make the viz rule depend on $(BIN)/$(HL_TARGET)-trace_all/bilateral_grid.mp4, and make the mp4 rules depend on filter, instead of filter_viz?

Reply:
Not without more work, because filter_viz also builds with -DNO_AUTO_SCHEDULE.